xc/xenctrlext.ml

```diff
@@ -83,5 +83,36 @@ external numainfo : handle -> numainfo = "stub_xenctrlext_numainfo"
 
 external cputopoinfo : handle -> cputopo array = "stub_xenctrlext_cputopoinfo"
 
-external xc_get_msr_arch_caps : handle -> int64
-  = "stub_xenctrlext_get_msr_arch_caps"
+let string_of_leaf v =
+  Printf.sprintf "%08Lx:%08Lx->%08Lx:%08Lx:%08Lx:%08Lx"
+  v.leaf v.subleaf v.a v.b v.c v.d
+
+let leaf_of_string s =
+  Scanf.sscanf s "%08Lx:%08Lx->%08Lx:%08Lx:%08Lx:%08Lx" @@
```

Comment:
Each of the captured fields enforce that at most are 8 characters long, but are not enforced to be at least 8 characters long.

If this is trusted input then it's not a problem, otherwise it is.

See [here](https://github.com/xapi-project/xen-api-libs-transitional/pull/103/commits/e0240349da8dea2c7b1900a9f327ff52fcc42cd7#diff-ac9d50fad3b608cc0a16868120b4f0df29345124b4265d16cd6370ab3cd636feR84) for a similar situation when parsing uuids with scanf

Reply:
The input here will be stored/retrieved from the XAPI DB, and can be set by VM role admin (eventually), perhaps I should remove the length specifiers from the format so we don't give the illusion that it validates the length. I'll see if I can add a unit test to see what happens on malformed input (on malformed input we should raise an exception that prevents the VM from starting, instead of starting it based on truncated featureset)